geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledSelect.java

```diff
@@ -925,79 +917,89 @@ private SelectResults applyProjectionOnCollection(SelectResults resultSet,
+    // If single attribute is present without alias name, then create ResultSet.
+    // Else if more than on attribute or single attribute with alias is present then return a
+    // StructSet.
+    // Create StructSet which will contain root objects of all iterators in from clause.
 
     ObjectType elementType = this.cachedElementTypeForOrderBy != null
         ? this.cachedElementTypeForOrderBy : prepareResultType(context);
     SelectResults results;
-    if (this.distinct || !this.count) {
-      if (this.orderByAttrs != null) {
-        boolean nullValuesAtStart = !((CompiledSortCriterion) orderByAttrs.get(0)).getCriterion();
-        if (elementType.isStructType()) {
-          if (ignoreOrderBy) {
-            results = this.distinct ? new LinkedStructSet((StructTypeImpl) elementType)
-                : new SortedResultsBag(elementType, nullValuesAtStart);
 
-          } else {
-            OrderByComparator comparator = this.hasUnmappedOrderByCols
-                ? new OrderByComparatorUnmapped(this.orderByAttrs, (StructTypeImpl) elementType,
-                    context)
-                : new OrderByComparator(this.orderByAttrs, (StructTypeImpl) elementType, context);
-            results = this.distinct ? new SortedStructSet(comparator, (StructTypeImpl) elementType)
-                : new SortedStructBag(comparator, (StructType) elementType, nullValuesAtStart);
+    if (!this.distinct && this.count) {
+      // Shobhit: If it's a 'COUNT' query and no End processing required Like for 'DISTINCT'
+      // we can directly keep count in ResultSet and ResultBag is good enough for that.
+      countStartQueryResult = 0;
+      return new ResultsBag(new ObjectTypeImpl(Integer.class), 1, context.getCachePerfStats());
+    }
 
-          }
-        } else {
-          if (ignoreOrderBy) {
-            results =
-                this.distinct ? new LinkedResultSet() : new SortedResultsBag(nullValuesAtStart);
+    // Potential edge-case: Could this be non-null but empty?
+    boolean nullValuesAtStart = orderByAttrs != null && !orderByAttrs.get(0).getCriterion();
+    OrderByComparator comparator;
+    switch (convertToStringCase(elementType, ignoreOrderBy)) {
+      case "UNORDERED DISTINCT STRUCT":
+        return new StructSet((StructType) elementType);
+      case "UNORDERED DISTINCT RESULTS":
+        return new ResultsSet(elementType);
+      case "UNORDERED INDISTINCT STRUCT":
+        return new StructBag((StructType) elementType, context.getCachePerfStats());
+      case "UNORDERED INDISTINCT RESULTS":
+        return new ResultsBag(elementType, context.getCachePerfStats());
+
+      case "ORDERED DISTINCT STRUCT IGNORED":
+        return new LinkedStructSet((StructTypeImpl) elementType);
+      case "ORDERED INDISTINCT STRUCT IGNORED":
+        return new SortedResultsBag(elementType, nullValuesAtStart);
+      case "ORDERED DISTINCT STRUCT UNIGNORED":
```

Comment:
Good suggestion - thanks! @PurelyApplied and I worked together to get rid of the Strings, and we're now using enums. You should be able to see these changes starting at around line 920 of `CompiledSelect` of the latest commit. We're using a naming convention that I think follows your suggestion (instead of `UNORDERED_DISTINCT_STRUCT_IGNORED`, for example, we just use `UNORDERED_DISTINCT_STRUCT`).